tools/introspection/source/introspection_app.cpp

```diff
@@ -699,36 +697,31 @@ void IntrospectionApp::runIntrospection(const iox::units::Duration updatePeriodM
         {
             prettyPrint("### MemPool Status ###\n\n", PrettyOptions::highlight);
 
-            const void* rawMempoolSample{nullptr};
+            bool hasMoreSamples{true};
 
-            while (!rawMempoolSample)
+            do
             {
-                memPoolSubscriber.getChunk(&rawMempoolSample);
-            }
-
-            const MemPoolIntrospectionInfoContainer* mempoolSample =
-                static_cast<const MemPoolIntrospectionInfoContainer*>(rawMempoolSample);
-
-            if (mempoolSample->empty())
-            {
-                prettyPrint("Waiting for mempool introspection data ...\n");
-            }
-            else
-            {
-                for (const auto& i : *mempoolSample)
-                {
-                    printMemPoolInfo(i);
-                }
-            }
-
-            memPoolSubscriber.releaseChunk(rawMempoolSample);
+                memPoolSubscriber.take()
+                    .and_then([&](iox::cxx::optional<iox::popo::Sample<const void>>& sample) {
+                        const MemPoolIntrospectionInfoContainer* receivedSample =
+                            static_cast<const MemPoolIntrospectionInfoContainer*>(sample->get());
+                        for (const auto& i : *receivedSample)
+                        {
+                            printMemPoolInfo(i);
+                        }
+                    })
+                    .if_empty([&] {
+                        hasMoreSamples = false;
+                        prettyPrint("Waiting for mempool introspection data ...\n");
+                    });
+            } while (!hasMoreSamples);
```

Comment:
this is the same problem budrus mentioned in a comment in `iceoryx.cpp`. The cache size is 1 and we just wait till the first sample arrives. If it's not immediately there, there is an infinite loop. I think inverting the logic with `bool sampleReceived{false}` works better. Nevertheless, the whole introspection app needs to be refactored or rewritten in python once we have python bindings ... btw. can we also get some `Go` bindings ;)

Reply:
Reworked loop!